New medals and medal code reorganization

Xhuis opened the pull request:
:cl: Xhuis
add: Several new achievements have been added.
/:cl:
- Medal code has been moved to `__HELPERS/medals.dm`.
- Individual medals have been moved to `__DEFINES/medals.dm`.
- Medals now have different announcement options, including announcing to nobody (silent), announcing to the earner, announcing to the earner and admins, and announcing to everyone.
- Refactored megafauna medal code and gave each megafauna its own medal.
- Added a bunch of new medals, including refactors.
  - **You'll Shoot Your Eye Out:** Kill every watcher on the map.
  - **David:** Kill every goliath on the map.
  - **Begone, Devils:** Kill every legion on the map.
  - **Peace Amongst Ashes:** Kill every necropolis tendril on the map.
  - **Slayer:** Kill a boss.
  - **The Necropolis Awaits:** Kill Legion.
  - **All out of Bubblegum:** Kill Bubblegum.
  - **Bullet Hell:** Kill a colossus.
  - **Science Fiction:** Kill an ash drake.
  - **For the Honkmother:** Enter a HONK mech as a clown.
  - **Jackpot:** Win a pulse rifle from an arcade machine.
  - **Witness to Doomsday:** Witness a clash between Nar-Sie and Ratvar.

Because of how medal code works, I'm not able to test this!
@Shadowlight213 You seem to be our lead medal guy. Are you okay with this?

Commit:
New medals and medal code reorganization

Durkel commented:
Do we really need achievements?

anconfuzedrock commented:
wait what the fuck are medals

Xhuis commented:
We just added code for achievements/medals and gave them to megafauna. I don't see a reason to ignore it.

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_ratvar.dm:
```diff
@@ -251,6 +254,8 @@
 		N.clashing = TRUE
 		clash_of_the_titans(N) //IT'S TIME FOR THE BATTLE OF THE AGES
+		for(var/mob/living/L in player_list)
+			UnlockMedal(L.client, MEDAL_GOD_CLASH)
```
I think you fucked up

Xhuis replied:
Misplaced syntax, gotcha

ChangelingRain commented on code/modules/power/singularity/narsie.dm:
```diff
@@ -34,6 +34,10 @@
+	for(var/datum/mind/M in all_clockwork_mobs)
+		var/mob/living/L = M.current
+		if(!L.stat && L.z == z && L.client)
+			UnlockMedal(MEDAL_SUMMON_RATVAR, L.client)
```
I think you really fucked up

Xhuis replied:
Eh? How so?

RemieRichards replied:
@Xhuis this is Narsie.

MMMiracles commented:
Nothing wrong with giving players a goal to strive for.

Commit:
Syntax fixes, removes Bag of Holding achievement

Xhuis commented:
* Fixed some issues with summoning achievements.
* Removed the only achievement that you might get banned for achieving.

Commit:
Adds a medal for entering a HONK as a clown

Xhuis commented:
* Added "For the Honkmother", awarded by entering a HONK mech as a clown.

KorPhaeron commented on code/__DEFINES/medals.dm:
```diff
@@ -0,0 +1,24 @@
+
+//These medals are feats of strength, awarded through difficult tasks or incredible endurance.
+#define MEDAL_LEGION "The Necropolis Awaits" //Kill Legion.
+#define MEDAL_BUBBLEGUM "All out of Bubblegum" //Kill Bubblegum.
```
I dont think you can rename these

Xhuis replied:
Is that a request or you saying that you don't actually know?

KorPhaeron replied:
I mean this will break the current medals

@shadowlight213

Shadowlight213 replied:
I'm not completely sure actually. Unlike the scoreboard the page does seem to have an option for renaming medals.
I'll run some tests.

Shadowlight213 replied:
@KorPhaeron @Xhuis I tested it. Changing the name of the medal via the interface works ok. It doesn't remove the medals from players who already earned them.
The scores are the ones that you can't change the names of.

lzimann commented:
will i lose my current achievements after this get merged

KorPhaeron commented:
yes

PKPenguin321 commented:
Aw man what the heck

This will super conflict with my meteor medal PR and also reset all our achievements :(

ike709 commented:
You should add a medal that only an admin can award. Except I don't know what you would call it or what it'd be for.

Xhuis commented:
@PKPenguin321 It shouldn't conflict at all. Achievements _might_ be reset unless I figure out how to prevent it. It might add in a workaround that checks for the old achievements for a while.

PKPenguin321 commented:
ah okay that's good
also i'm pretty sure @Shadowlight213 needs to approve of these and make appropriate changes on the hub page

lzimann commented:
i just don't want to lose my medals i don't care about the rest

KorPhaeron commented:
What if we just don't arbitrarily rename achivements since it will reset the scoreboard

Commit:
Optimizes a lot of code

Xhuis commented:
* Re-added the achievement for killing any boss.
* Refactored Exterminator achievements into separate achievements. Scores are unaffected.
* Made admin-spawned Nar-Sie and Ratvar not grant achievements.
* Moved the `elimination` variable to Legion instead of at the base megafauna type.
* Made a lot of code less spaghetti.

As @Shadowlight213 said earlier, scores aren't reset and he just has to rename the achievements and progress will carry over for everyone.

Commit:
Removes some debug code

lordpidey commented:
Hey, there's a couple more achievements I'd like to see you add.

Ascend as a sling
Ascend as a devil
Fail the die gloriously objective.
Win a round as a HoG deity.
Survive and be on shuttle at end of round when you are a traitor's target.

PKPenguin321 commented on code/__DEFINES/medals.dm:
```diff
@@ -7,6 +7,12 @@
 //These medals are feats of strength, awarded through difficult tasks or incredible endurance.
+#define MEDAL_WATCHER_CLEANER "You'll Shoot Your Eye Out" //Kill every watcher on the map.
+#define MEDAL_GOLIATH_CLEANER "David" //Kill every goliath on the map.
+#define MEDAL_LEGION_CLEANER "Begone, Devils" //Kill every watcher on the map.
```
inaccurate comment

PKPenguin321 commented:
why are we renaming the achievements

Xhuis commented:
Because "Legion Killer" isn't as cool as "The Necropolis Awaits".

KorPhaeron commented:
It's just a stat tracking board for an atmos sim it doesn't need to be """cool""" it needs to be readable

MMMiracles commented:
if an atmospheric simulator is going to have a mini-hell with giant hell-skull boss it should have a cool-sounding achievement.

optimumtact commented on code/__HELPERS/medals.dm:
```diff
@@ -0,0 +1,119 @@
+						player << "<span class='noticealien'>Achievement unlocked:</span> <span class='alertalien'>[medal]</span><span class='noticealien'>!</span>"
+						message_admins("Player [player.ckey] unlocked medal [medal]!")
+					if(ANNOUNCE_TO_EVERYONE)
+						player << "<span class='noticealien'>[player.ckey] has earned the</span> <span class='alertalien'>[medal]</span> <span class='noticealien'>achievement!</span>"s
```
floating s on the end of this line

Xhuis replied:
Thanks

KorPhaeron commented:
Xhuis: 2. I can't test anything so I'm relying on you as a maintainer to make sure that it's mergeable

There has to be some way of testing this other than seeing if it breaks everything

Xhuis commented:
If there is, I have no idea. I don't have access to the medal hub to edit/add/remove them, so I can't test to see if these new medals actually exist.

Commit:
Now it compiles!

Commit:
Adds a new achievement

Xhuis commented:
* Added the "One Job..." achievement, awarded for failing your martyr objective. Dummy.

KorPhaeron commented:
You could set up your own test hub

Xhuis commented:
Any idea how?

Commit:
I swear I know how to code

KorPhaeron commented:
Ask @Shadowlight213

Shadowlight213 commented:
@Xhuis @KorPhaeron 
The medals don't exist until I add them to the hub or change the name of an existing medal.
The medal setting and clearing works off of strings. Sending an incorrect string to the hub won't break existing medals.
The worst case scenario I can think of would be it returning null and activating the safety switch. I can run tests to see if that happens anyways.
Anyways. If you really want to, I can help you setup a hub or just make one myself and give you access.

Shadowlight213 commented on code/__HELPERS/medals.dm:
```diff
@@ -0,0 +1,119 @@
+				global.medals_enabled = FALSE
+				log_game("MEDAL ERROR: Could not contact hub to award medal:[medal] player:[player.ckey]")
+				message_admins("Error! Failed to contact hub to award [medal] medal to [player.ckey]!")
+			else
```
This is wrong. I had the else if(result) there for a reason.
If the medal is already unlocked, the hub returns 0. With this else, it will announce it every time.

Shadowlight213 commented on code/__HELPERS/medals.dm:
```diff
@@ -0,0 +1,119 @@
+			message_admins("Error! Failed to contact hub to clear [medal] medal for [player.ckey]!")
+		else if(result)
+			message_admins("Medal: [medal] removed for [player.ckey]")
+		else
```
See above. I have those if statements setup how i had them for a reason.
Hub returns 0 if the medal wasn't awarded in this case.

Xhuis replied:
Thanks, will update.

Shadowlight213 commented on code/game/gamemodes/clock_cult/clock_ratvar.dm:
```diff
@@ -251,6 +255,9 @@
 		N.clashing = TRUE
 		clash_of_the_titans(N) //IT'S TIME FOR THE BATTLE OF THE AGES
+		if(!admin_spawned)
+			for(var/mob/living/L in player_list)
```
I feel like this needs some sort of limit.
A drone on the derelict or especially people playing ctf shouldn't be included.

Xhuis replied:
I'll add a same z-level check.

Shadowlight213 commented on code/__HELPERS/medals.dm:
```diff
@@ -0,0 +1,119 @@
+		else if(result)
+			message_admins("Medal: [medal] removed for [player.ckey]")
+		else
+			log_game("Player [player.ckey] unlocked medal [medal].")
```
I don't think this is what you meant.

Xhuis replied:
It is. What did you think I did?

Shadowlight213 replied:
That's a log for UNlocking a medal in the LOCK proc. Where the rest of the stuff says lock as well.

Xhuis commented:
Setting up a hub would be helpful for testing. Is it separate from the game's pager entry? Something I can just do in 5 minutes?

Shadowlight213 commented:
@Xhuis It normally is, but the way i have the proc setup, it uses the destination and password supplied in the config instead.
To make a hub entry, login to the byond website, hover over your name, then click publish games in the dropdown them add a new game on the page it takes you to. Shouldn't take more than 5 min.

Xhuis commented:
Thanks bud

Xhuis commented:
Just tested the unpushed commit. Everything worked pretty much as intended, although I haven't tried scores yet since those are untouched.

Commit:
Fixes some incorrect coding

Xhuis commented:
* Fixed some syntax that could cause runtimes. (Thanks @Shadowlight213!)
* Witness to Doomsday is now unlocked as the clash starts, rather than when it ends. In addition, only mobs on the clash's z-level unlock the achievement.

I'm gonna tentatively say that this is merge-ready. All testing on my end has come back positive, but you might want to make sure on yours just to make sure that this doesn't break anything.

lzimann commented:
why arent the medals name the same as it is right now, i don't want to lose it

ChangelingRain commented on code/game/gamemodes/clock_cult/clock_ratvar.dm:
```diff
@@ -250,6 +254,10 @@
 		if(N.clashing)
 			continue
 		N.clashing = TRUE
+		if(!admin_spawned)
```
Make sure the nar-sie is also not admin-spawned.

Xhuis replied:
Will do that tomorrow morning

Xhuis commented:
Izimann, have you been paying attention at all? They're _not_ the same. Once this gets merged they'll be renamed so you won't lose your progress.

Core0verload commented:
There are some fun medals I found in BS12-Luna code:
**That Tasted Funny**
(for eating clown meat or clown meat burger)

**DUI**
(for being drunk on a mulebot as it runs someone over)

**Survivor** 
What do you think?
(for surviving the round)

**Slave to the Overmind**
You are a pawn of the AI.
(for being slaved to AI as a cyborg)

Kromgar commented:
**Rend and Tear**
You have slain all the beasts of Hell. (Colossi, Bubblegum, Legion, and Drake)


Unlocks: Doomguy Mining Rig skin

Xhuis commented:
There's been some pretty big family issues going on in my life, so don't expect any updates on this for a bit. I'm not abandoning it, but it'll be stale for a few days.

optimumtact commented:
Copy that

Shadowlight213 commented:
Needs a "least robust person ever" medal for dying to a temp gun.

ike709 commented:
Should also add a medal for being an ERT member

KorPhaeron commented on code/__DEFINES/medals.dm:
```diff
@@ -0,0 +1,40 @@
+#define MEDAL_HIJACK "Crashing This Station..." //Hijack the emergency shuttle.
+#define MEDAL_SUMMON_RATVAR "Dawn Rising" //As a clockwork cultist, summon Ratvar.
+#define MEDAL_SUMMON_NARSIE "Darkness Falling" //As a cultist, summon Nar-Sie.
+#define MEDAL_NUCLEAR "Radioactive, Radioactive" //As a nuclear operative, survive from start to finish and nuke the station.
```
It's already hard enough to get an op team to do fun things and break the meta without super greentext for it

KorPhaeron commented on code/__DEFINES/medals.dm:
```diff
@@ -0,0 +1,40 @@
+//These medals are mundane or hold no real significance.
+#define TEST_MEDAL "Hello World!" //Unachieveable.
+
+#define MEDAL_NEW_PLAYER "Fresh Meat" //Log in for the first time.
```
This is just going to spam and clog up the medal page

Shadowlight213 replied:
agreed

KorPhaeron commented on code/__DEFINES/medals.dm:
```diff
@@ -0,0 +1,40 @@
+#define TEST_MEDAL "Hello World!" //Unachieveable.
+
+#define MEDAL_NEW_PLAYER "Fresh Meat" //Log in for the first time.
+#define MEDAL_FIRST_DEATH "Remember This Moment" //Die for the first time.
```
So is this one. Don't know why we need "started the game" style achievements

KorPhaeron commented on code/__DEFINES/medals.dm:
```diff
@@ -0,0 +1,40 @@
+
+//These medals are games of chance, either through extreme luck or extreme bravery.
+#define MEDAL_ARCADE_RNG "Jackpot" //Hit the one-in-a-million and win a pulse rifle from an arcade machine.
+#define MEDAL_GOD_CLASH "Witness to Doomsday" //Witness a clash between Nar-Sie and Ratvar.
```
Adding a medal for a mode that doesnt exist is a bit odd

optimumtact replied:
Admins can create these situations

Shadowlight213 replied:
True. An admin could trigger both a clock and blood cult and it wouldn't cause the spawned narsie or ratvar to be invalidated.

PKPenguin321 commented:
medal for killing a blob when

Shadowlight213 commented:
I was talking with kor earlier, and we kinda agreed that this is way too many medals and for far too generic things.
Medals should be for something rare, not how modern games make achievements for beating level 1 or opening the inventory.
Stuff like completing your objective should not be made medals.
Meteor examining was kinda pushing it tbh, but it's pretty hard to pull off.

@PKPenguin321 No. I do not have any plans to add more things to the leaderboard. It's not very well formatted to handle tons of columns and part of it's function is providing accurate feedback data for kor.

Xhuis commented:
Point taken. I'll prune a bunch of the ones that seem easy to get but a few will stay, like the clown one and the god clash one.

KorPhaeron commented:
Clown one is pretty good yeah

PKPenguin321 commented:
>Meteor examining was kinda pushing it tbh, but it's pretty hard to pull off.

nobody's managed to do it yet and i've seen at least 1 person die trying to examine a meteor so i'd say it's a pretty good achievement :smirk:

Shadowlight213 commented:
Ayy. Yeah I'll agree on that I guess.

optimumtact commented:
Could you finish this soon please?

Xhuis commented:
Sure, I'll get some work done today.

Commit:
Removes some medals

Xhuis commented:
I cropped a few of the medals you suggested. OP has been updated.

Commit:
:(

Xhuis commented:
Okay. I _actually_ removed the medals now.

Xhuis commented:
Going to close this for now, as I want to work on other things. I'll probably bring it back this weekend.
